blog/2023-02-12-conditional-export.md

```diff
@@ -0,0 +1,137 @@
+
+## The crux
+
+The traditional way of doing this is to build two versions of the package (`dev` with the dev tools included and `prod` without them) and use it in `index.js` (referenced from package.json) with conditional `require` based on `process.env.NODE_ENV`.
+
+In our case this looks something like this:
+
+```ts
+// package.json
+{
+  ...
+  "main": "index.js"
+  ...
+}
+```
+
+```json
+// index.js
+module.exports =
+  process.env.NODE_ENV === 'production'
+    ? require('./dist/bundle.prod')
+    : require('./dist/bundle.dev');
+```
+
+This looks straightforward, right? It's not as easy as you might think.
+
+### How does it work under the hood?
+
+The important thing to keep in mind here is that without any optimizations the bundler would include both files because the condition result is unknown at the build time.
+
+The trick is that it's a non-written standard that the conventional build setups replace `process.env.NODE_ENV` with `production` when they are building the app for production and so after then the optimizer has a code that looks like this:
+
+```ts
+module.exports =
+  'production' === 'production'
+    ? require('./dist/bundle.prod')
+    : require('./dist/bundle.dev');
+```
+
+It's now pretty clear for the optimizer that the `dev` branch can be eliminated, as it will never be executed.
+
+### Do we need to use require?
+
+It is tempting to change the code to es module style, but as the `import` statements can't be used conditionally. We would need to do something like this:
+
+```ts
+export * from './dist/bundle.prod';
+import * as devBundle from './dist/bundle.dev';
+
+if (process.env.NODE_ENV === 'production') {
+  export const DevTools = devBundle.DevTools;
+} else {
+  export const DevTools = <dummy function>
+}
+```
+
+The problem is that `devBundle` is imported unconditionally and it's only the export part that is conditional.
+The module still can have some side effects and in that case, the optimizer can't remove it automatically because that could change the behavior.
+
+There are ways how to tell the compiler to consider it side-effect free, but that's quite heavily dependent on the capability of the optimizer that is used on the user side. It's not reliable.
+
+### What if we can't use `require`?
+
+Modern tools like `VITE` completely shifted to es modules and will give you an error if you use `require` in the code.
+
+What can we do to conditionally switch between the `prod` and `dev` bundles? Luckily there is a way.
+
+#### Magic with package.json
+
+There is a standard field in package.json to define [conditional exports](https://nodejs.org/api/packages.html#conditional-exports). It is used by node js to distinguish between es module and cjs bundles like this:
+
+```json
+{
+  "exports": {
+    "import": "./dist/bundle.esm.js",
+    "require": "./dist/bundle.cjs.js"
+  }
+}
+```
+
+Modern build tools utilize this feature by using semi-standard fields "production" and "development", so we can do:
```

Comment:
Can you specify which modern tools you have in mind specifically.

Reply:
I'm mentioning VITE in few sentences before, I think it's not necessary to write it again